crates/client/eth/src/l1_messaging.rs

```diff
@@ -0,0 +1,497 @@
+        };
+
+        let _ = contract.setIsCanceled(false).send().await;
+        // Send a Event and wait for processing, Panic if fail
+        let _ = contract.fireEvent().send().await.expect("Failed to fire event");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+
+        // Assert that event was caught by the worker with correct data
+        // TODO: Maybe add some more assert
+        assert!(logs_contain("fromAddress: 0xae0ee0a63a2ce6baeeffe56e7714fb4efe48d419"));
+
+        // Assert the tx hash computed by the worker is correct
+        assert!(logs_contain(
+            format!("event hash : {:?}", contract.getL1ToL2MsgHash().call().await.expect("failed to get hash")._0)
+                .as_str()
+        ));
+
+        // TODO : Assert that the tx has been included in the mempool
+
+        // Assert that the event is well stored in db
+        let last_block =
+            db.backend().messaging_last_synced_l1_block_with_event().expect("failed to retrieve block").unwrap();
+        assert_ne!(last_block.block_number, 0);
+
+        // TODO : Assert that the tx was correctly executed
+
+        // Explicitly cancel the listen task, else it would be running in the background
+        worker_handle.abort();
+    }
+
+    /// Test the workflow of l1 -> l2 messaging with duplicate event
+    ///
+    /// This test performs the following steps:
+    /// 1. Sets up test environemment
+    /// 2. Starts worker
+    /// 3. Fires a Message event from the dummy contract
+    /// 4. Waits for event to be processed
+    /// 5. Assert that the event is well stored in db
+    /// 6. Fires a Message with the same event from the dummy contract
+    /// 7. Assert that the last event stored is the first one
+    #[rstest]
+    #[traced_test]
+    #[tokio::test]
+    async fn e2e_test_already_processed_event(#[future] setup_test_env: TestRunner) {
+        let TestRunner { chain_config, db_service: db, dummy_contract: contract, eth_client, anvil: _anvil } =
+            setup_test_env.await;
+
+        // Start worker
+        let worker_handle = {
+            let db = Arc::clone(&db);
+            tokio::spawn(async move { sync(db.backend(), &eth_client, &chain_config.chain_id).await })
+        };
+
+        let _ = contract.setIsCanceled(false).send().await;
+        let _ = contract.fireEvent().send().await.expect("Failed to fire event");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+        let last_block =
+            db.backend().messaging_last_synced_l1_block_with_event().expect("failed to retrieve block").unwrap();
+        assert_ne!(last_block.block_number, 0);
+
+        // Send the event a second time
+        let _ = contract.fireEvent().send().await.expect("Failed to fire event");
+        tokio::time::sleep(Duration::from_secs(5)).await;
+        // Assert that the last event in db is still the same as it is already processed (same nonce)
+        assert_eq!(
```

Comment:
should we also check nonce?

Reply:
done

Reply:
is this pending?

Reply:
no no its live. Maybe you had same bug as me where github doesn't show change from all commit in te "File changed" tab

Reply:
oh nvm, I guess I was checking at the wrong place, sorry